### ISSUE TITLE
Teach chain certificates to SSL server endpoints string syntax

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6499]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)| @hynek reported|
|-|-|
|Trac ID|trac#6499|
|Type|enhancement|
|Created|2013-04-30 08:57:25Z|
|Branch|https://github.com/twisted/twisted/tree/ssl-endpoint-string-chain-certificate-6499|

Now we have chain certificates, `twisted.internet.endpoints._parseSSL` should be able to handle them too.

Attachments:

* [6499-chains-for-strports.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/8ea/8eaf166e0b79deeb28d02356cd9ff1fbb14fa6ee/2358832914f5ab3170d7f128277faa3ced957405.diff) (9351 bytes) - added by hynek on 2013-05-11 12:26:52Z - 
* [6499-chains-for-strports-v2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/8ea/8eaf166e0b79deeb28d02356cd9ff1fbb14fa6ee/f9a522a07c4e37cad816f57bbe17159179af4009.diff) (9597 bytes) - added by hynek on 2013-05-11 12:45:31Z - Topfile + grammar.
* [6499-chains-for-strports-v3.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/8ea/8eaf166e0b79deeb28d02356cd9ff1fbb14fa6ee/76fc642f84300b54959d70e08a04745c036b0825.diff) (11792 bytes) - added by hynek on 2013-05-11 15:11:48Z - Add endpoints documentation.
* [6499-chains-for-strports-v4.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/8ea/8eaf166e0b79deeb28d02356cd9ff1fbb14fa6ee/8e33d27bbd204b8a8fa15bace209658eaaa909b8.diff) (11070 bytes) - added by hynek on 2013-05-27 08:02:16Z - Updates as per zooko’s & exarkun’s feedback.
* [6499-chains-for-strports-v5.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/8ea/8eaf166e0b79deeb28d02356cd9ff1fbb14fa6ee/394866983c1c08668fa6d67f89f6bccfe92bc31a.diff) (11084 bytes) - added by hynek on 2013-05-27 09:15:11Z - Fix outdated comment.
* [fix-ci-errors.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/8ea/8eaf166e0b79deeb28d02356cd9ff1fbb14fa6ee/e2693aa5dde61b74e1e0b41325f3be873d843855.patch) (1499 bytes) - added by hynek on 2013-06-14 19:33:39Z - 
* [fix-python3-and-ssl.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/8ea/8eaf166e0b79deeb28d02356cd9ff1fbb14fa6ee/8adf8ae2c1635a1aeaed9ee6f39c816ed847f4c0.patch) (1623 bytes) - added by hynek on 2013-06-16 07:53:43Z - Fix Python 3.3 & pyOpenSSL regression.
* [fix-tests-for-windows.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/8ea/8eaf166e0b79deeb28d02356cd9ff1fbb14fa6ee/d576ed07bbe101d30c13526439c9dfe8f6044d6b.patch) (600 bytes) - added by hynek on 2013-06-16 19:59:21Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__6499 6499
type__enhancement enhancement
reporter__hynek hynek
priority__normal normal
milestone__ 
branch__branches_ssl_endpoint_string_chain_certificate_6499 branches/ssl-endpoint-string-chain-certificate-6499
branch_author__tomprince tomprince
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1367312245000000 1367312245000000
changetime__1371855997000000 1371855997000000
version__None None
owner__tom_prince tom.prince
cc__hynek cc__glyph cc__zooko@...
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

As part of this, it would be great to add an example demonstrating the use of chain certificates to the SSL howto.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_3"></a>|@hynek commented|
|-|-|

So while the code here is rather trivial, testing it properly is rather daunting (like everything SSL-related :-/).

Ideally we’d add a full fake CA to the testing code base so we have both more keys and certificates to play with but also to be able to generate new ones on demand. For example I need a file with multiple certificates right now. The strports code on the other hand uses one certificate file for everything – and doesn’t even bother to check whether the file is correctly imported (cf [twisted/internet/test/test_endpoints.py#L1798](twisted/internet/test/test_endpoints.py#L1798) ).

Any opinions on how to proceed here? I’m stuck here until this gets resolved but I don’t feel like press ahead without consultation.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Try not to be stuck for too long.  If you need to generate 3 or for keys and certificates for some new tests and check them in, fine.  Maybe it's not the totally ideal solution, but it lets you make progress so it's not all bad.  Afterwards, if you want to make a pass over all the certificate material in Twisted's test suite and organize it a little bit, that wouldn't hurt either.

I _do_ want to recommend against generating new certificates _during_ test runs.  It's slow.  I prefer fast tests and pre-generated crypto objects.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_5"></a>|@hynek set owner to @glyph|
|-|-|

First regarding the SSL howto: I haven’t done anything yet since it’s using `Certificate.options()` which is a very limited API not allowing chain files. I’m afraid we’ll need to close [#6361](https://github.com/twisted/twisted/issues/6361) first – or rewrite lots of the howto.

As for test data, I decided to simply concatenate the existing certificates which gave me nice verification possibilities.  I also needed a file that definitely doesn’t contain any certificates, so I went for `__init__.py`. If you want me to add some lorem ipsum instead I can do that too…

If you want to play with it, I’ve added a second example in https://gist.github.com/hynek/4095c70615a223ecc1df that uses strports.

As per his mail on the mailing list, I’m assigning the review to glyph but I’m sure he’ll allow others to look into the code too. ;)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst commented|
|-|-|

The endpoints howto should document this syntax in the SSL servers section, especially since chain certs are basically a necessity.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_7"></a>|@hynek commented|
|-|-|

FTR, I’ve opened a ticket at [#6513](https://github.com/twisted/twisted/issues/6513) for client chain file string syntax.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)|@zooko set owner to @zooko<br>@zooko set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_9"></a>|@zooko set owner to @hynek<br>@zooko set status to `new`|
|-|-|

This is my review of [6499-chains-for-strports-v3.diff](None/6499-chains-for-strports-v3.diff) .

Replace "trustworthy" with "root" in "from a trustworthy CA to the one that signed you certificate" ("trustworthy" is not the technical term for this, and is also inaccurate). Replace "you" with "your". While you're at it, replace "the chain of trust" with "the chain".

Also in the docstring, replace "trustworthy" with "root" and remove "of trust".

Why is it "notACertQuotedPathName" instead of "escapedNotACertPathName", like the others? (The others are "escapedChainPathName", "escapedCAsPathName", and "escapedPEMPathName".)

It would make me feel _slightly_ better if the test for "are there no certs here?" was moved down one line to be a test of ```chainCertificates``` instead of a test of ```matches```. I can't come up with a failure scenario which would justify this, but I just feel better checking that the fully-parsed objects are present rather than checking that the half-parsed strings are present. (Note that this half-parsing represented by the ```re.findall``` is the first parsing of certs in Twisted or pyOpenSSL (excluding test code). Hopefully also the last.)

Okay, otherwise this patch looks good to me! It is good that it includes a test of the exception raised when the file doesn't contain any certs. Thank you for writing this patch! I really need this fix. ☺

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

Thanks hynek, zooko.

In addition to the comments above, please consider:

  1. doc/core/howto/endpoints.xhtml
    1. Reflowing prose makes for super difficult reviews (presumably due to shortcomings of our tools). :(  Some people have suggested that one sentence per line in prose is a good compromise.  Since the text here is already word-wrapped, I would suggest just making edits *without* rewrapping it.
    1. Instead of "certKey, privateKey, extraCertChain and sslmethod." please write "certKey, privateKey, extraCertChain, and sslmethod."
  1. twisted/internet/endpoints.py
    1. The type documentation for the new parameter, `extraCertChain`, has two issues.  
       1. It should be `L{}` instead of `C{}` (I know the nearby parameters use `C{}`, they date from an earlier standard).  
       1. It should not be `str` now that `str` has become ambiguous.  All new documentation (and eventually all old documentation) should be written using `bytes` or `unicode`.  There may be some exceptions where `str` is still appropriate, but you can be pretty sure they won't involve user input.
  1. twisted/internet/test/test_endpoints.py
    1. Files not containing certificates are cheap.  `self.mktemp()` nearly gives you one, even.  I think using `__init__.py` for `notACertPath` is a little too clever.  Particular for a file with unit tests for SSL code, including certificate handling, who's to say there will never be a certificate in this file?
    1. I'd like to see the new chain-related code and assertions in `ServerStringTests.test_ssl` moved into new test methods.  The fewer assertions in a test method, the better (at least, until you get to 1).
  1. twisted/topfiles/6499.feature
    1. Maybe we'll never be free of the curse of ''strports''... but we can try.  This branch is concerned with "string endpoint descriptions" - `twisted.internet.endpoints.serverFromString`, not `twisted.application.strports`.  I suggest describing the feature as an improvement to "SSL server endpoint string descriptions" rather than "strports".

I like that zooko pointed out that this is the first instance of PEM parsing code anywhere in Twisted (or even pyOpenSSL).  That makes me somewhat uneasy.  It would be cool if we could find an API in OpenSSL to do this for us (but a cursory investigation hasn't revealed such an API to me).  I also briefly worried about the ".+?" in the regular expression, but I can't see how it could cause a real world problem.

Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_11"></a>|@hynek removed owner|
|-|-|

I hope I have incorporated all feedback both in word and spirit.

As for the docs:I have unwrapped lines that have been touched anyway, I hope that’s the right compromise and a good strep towards getting it like that everywhere eventually.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_12"></a>|@hynek commented|
|-|-|

Since I seem to have triggered some error in trac before ( http://glui.me/?i=3vnrs4xp6z2df4j/2013-05-27_at_10.10.26.png/ ) and apparently no notifies went out, I hereby ping it once more that I’ve updated the patch according to the feedback.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_13"></a>|@tomprince commented|
|-|-|

(In [38479]) Branching to ssl-endpoint-string-chain-certificate-6499.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_14"></a>|@tomprince commented|
|-|-|

(In [38480]) Apply 6499-chains-for-strports-v5.diff from hynek.

Refs: [#6499](https://github.com/twisted/twisted/issues/6499)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_15"></a>|@tomprince set owner to @hynek|
|-|-|

I started to look at this, but the last patch, which I applied to a branch doesn't appear to have all the code, and I'm not sure what else should be applied there. Can you attach a patch against the branch that has everything?

Also, there is a python3 [//buildbot.twistedmatrix.com/builders/python-3.3-tests/builds/857/steps/shell/logs/stdio build failure] and [//buildbot.twistedmatrix.com/builders/documentation/builds/3326/steps/api-documentation/logs/new%20pydoctor%20errors pydoctor error] (see [#6318](https://github.com/twisted/twisted/issues/6318) for some hints on `C{}` vs. `L{}`).

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_16"></a>|@hynek set owner to @tomprince|
|-|-|

I’ve addressed the build errors.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_17"></a>|@tomprince commented|
|-|-|

(In [38803]) Apply fix-ci-errors.patch from hynek.

Refs: [#6499](https://github.com/twisted/twisted/issues/6499)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_18"></a>|@tomprince set owner to @hynek|
|-|-|

1.Regarding parsing chain cert files, from a cursory perusal of openssl code, it appears that the way to read multiple certificates from a file is likely just to try and read them consecutively from the same bio. So please file tickets against pyopenssl and twisted to be able to support that.
   - Until that happens, I think the solution here is a reasonable compromise.
 2. Rather than creating a new in-tree file, it would be better to create the file on the fly (see [#6535](https://github.com/twisted/twisted/issues/6535)). But certainly use pregenerated data.
   - Since this is a minor point, and the existing tests use existing files, I'll merge as is, but please file a ticket for cleaning this all up.

Please merge and file the appropriate tickets.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_19"></a>|@hynek commented|
|-|-|

Not sure if you meant **me** with “please merge” and if so I’m flattered but I can’t. :)

As for your ticket requests:

 1. I’m not 100% sure what the contents of the tickets should be. “Add function to parse files with multiple certificates”? Is that something that belongs into Twisted at all?
 1. [#6578](https://github.com/twisted/twisted/issues/6578), I will tackle that soonish.

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)<a name="note_20"></a>|@ashfall set status to `closed`|
|-|-|

(In [38813]) Merge ssl-endpoint-string-chain-certificate-6499: Making SSL server endpoint string descriptions support the specification of chain certificates.

Author: hynek
Reviewers: zooko, exarkun, tomprince
Fixes: [#6499](https://github.com/twisted/twisted/issues/6499)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_21"></a>|@tomprince set status to `reopened`|
|-|-|

(In [38817]) Revert r38813: python 3.3 regression.

Reopens: [#6499](https://github.com/twisted/twisted/issues/6499)

```
Traceback (most recent call last):
  File "admin/run-python3-tests", line 30, in <module>
    unittest.main(module=None, argv=["run-python3-tests", "-v"] + testModules)
  File "/usr/lib/python3.3/unittest/main.py", line 124, in __init__
    self.parseArgs(argv)
  File "/usr/lib/python3.3/unittest/main.py", line 168, in parseArgs
    self.createTests()
  File "/usr/lib/python3.3/unittest/main.py", line 175, in createTests
    self.module)
  File "/usr/lib/python3.3/unittest/loader.py", line 137, in loadTestsFromNames
    suites = [self.loadTestsFromName(name, module) for name in names]
  File "/usr/lib/python3.3/unittest/loader.py", line 137, in <listcomp>
    suites = [self.loadTestsFromName(name, module) for name in names]
  File "/usr/lib/python3.3/unittest/loader.py", line 96, in loadTestsFromName
    module = __import__('.'.join(parts_copy))
  File "/var/lib/buildbot/buildarea/python-3.3-tests/Twisted/twisted/internet/test/test_endpoints.py", line 1780, in <module>
    class ServerStringTests(unittest.TestCase):
  File "/var/lib/buildbot/buildarea/python-3.3-tests/Twisted/twisted/internet/test/test_endpoints.py", line 1845, in ServerStringTests
    % (escapedPEMPathName,))
NameError: name 'escapedPEMPathName' is not defined
```

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_22"></a>|@hynek removed owner<br>@hynek set status to `new`|
|-|-|

This is embarrassing, I was under the impression that we don’t test Python 3.3 with SSL. Since the SSL-code is riddled with these path-names, I thought it’s enough to protect them like that. Oh well, there’s obviously more to it.

I made the code less clever and it passes regardless of the presence of PyOpenSSL on Python 3.3.

I’m sorry everyone.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_23"></a>|@radix commented|
|-|-|

I've applied the patch to the branch, so the branch should now be up to date.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_24"></a>|@hynek commented|
|-|-|

The error message of the XP buildbot is a bit cryptic but since the empty temporary file was unquoted, I’ll presume it’s the reason for it. Applying and buildbot-nudging would be appreciated to go for sure. :)

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_25"></a>|@hynek commented|
|-|-|

For the record, yes the previous patch fixed the SSL error in the XP buildbot.

The Python 3.3 errors that caused the revert have been addressed too.

Therefore this branch is ready for final review/merge again.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_26"></a>|@tomprince set owner to @tomprince<br>@tomprince set status to `assigned`|
|-|-|

The buildbots look happy, and it is good not having double interpolation on the string.

I'll merge this.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_27"></a>|@tomprince set status to `closed`|
|-|-|

(In [38882]) Merge ssl-endpoint-string-chain-certificate-6499: Make SSL server endpoint string descriptions support the specification of chain certificates.

Author: hynek
Reviewers: zooko, exarkun, tomprince
Fixes: [#6499](https://github.com/twisted/twisted/issues/6499)
